File: backend/app/schemas/campaign.py
from datetime import datetime

from pydantic import BaseModel, Field


class CampaignCreate(BaseModel):
    name: str = Field(max_length=255)
    original_subject: str = Field(max_length=500)
    original_body: str
    rotate_every_n: int = 5
    delay_min_sec: float = 5.0
    delay_max_sec: float = 30.0
    skip_validation: bool = False


class CampaignUpdate(BaseModel):
    name: str | None = None
    original_subject: str | None = None
    original_body: str | None = None
    rotate_every_n: int | None = None
    delay_min_sec: float | None = None
    delay_max_sec: float | None = None
    skip_validation: bool | None = None


class CampaignRead(BaseModel):
    id: int
    name: str
    original_subject: str
    original_body: str
    total_contacts: int
    valid_contacts: int
    sent_count: int
    failed_count: int
    status: str
    rotate_every_n: int
    delay_min_sec: float
    delay_max_sec: float
    skip_validation: bool
    started_at: datetime | None
    completed_at: datetime | None
    created_at: datetime

    model_config = {"from_attributes": True}


class CampaignStats(BaseModel):
    total_contacts: int
    valid_contacts: int
    sent_count: int
    failed_count: int
    invalid_count: int
    remaining: int
    progress_percent: float
    elapsed_seconds: float | None
    eta_seconds: float | None


class ValidationBreakdown(BaseModel):
    syntax_errors: int
    no_mx: int
    smtp_rejected: int
    disposable: int
    other: int


class CampaignReport(BaseModel):
    total_contacts: int
    valid_contacts: int
    invalid_contacts: int
    sent_count: int
    failed_count: int
    pending_count: int
    validation_breakdown: ValidationBreakdown

File: frontend/src/types/campaign.ts
export type CampaignStatus =
  | "draft"
  | "validating"
  | "generating"
  | "ready"
  | "running"
  | "paused"
  | "completed"
  | "failed";

export interface Campaign {
  id: number;
  name: string;
  original_subject: string;
  original_body: string;
  total_contacts: number;
  valid_contacts: number;
  sent_count: number;
  failed_count: number;
  status: CampaignStatus;
  rotate_every_n: number;
  delay_min_sec: number;
  delay_max_sec: number;
  skip_validation: boolean;
  started_at: string | null;
  completed_at: string | null;
  created_at: string;
}

export interface CampaignCreate {
  name: string;
  original_subject: string;
  original_body: string;
  rotate_every_n?: number;
  delay_min_sec?: number;
  delay_max_sec?: number;
  skip_validation?: boolean;
}

export interface CampaignStats {
  total_contacts: number;
  valid_contacts: number;
  sent_count: number;
  failed_count: number;
  invalid_count: number;
  remaining: number;
  progress_percent: number;
  elapsed_seconds: number | null;
  eta_seconds: number | null;
}

export interface ValidationBreakdown {
  syntax_errors: number;
  no_mx: number;
  smtp_rejected: number;
  disposable: number;
  other: number;
}

export interface CampaignReport {
  total_contacts: number;
  valid_contacts: number;
  invalid_contacts: number;
  sent_count: number;
  failed_count: number;
  pending_count: number;
  validation_breakdown: ValidationBreakdown;
}
